### PR TITLE
PLUGIN-81 - Splunk Streaming Source

### PR DESCRIPTION
**Pull request by rostyslavmorozov**
WIKI: https://wiki.cask.co/display/CE/Splunk+Plugin
JIRA: https://issues.cask.co/browse/PLUGIN-81

Included:
- Streaming Source

### CODE REVIEW

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+          long iteratorPosition = searchIterator.getIteratorPosition();
+          searchIterator = new SplunkSearchIterator(config, searchId, iteratorPosition, allRecordsCount);
+          searchIterator.setIteratorPosition(iteratorPosition);
+        }
+      }
+    } catch (Exception e) {
+      String errorMessage = "Exception while receiving messages from hubspot";
+      // Since it's top level method of thread, we need to log the exception or it will be unseen
+      LOG.error(errorMessage, e);
+      throw new RuntimeException(errorMessage, e);
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:83]: bad practice to log and re-throw exception. Won't CDAP log it?

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSource.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  public SplunkStreamingSource(SplunkStreamingSourceConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    config.validate(failureCollector);
+    config.validateConnection(failureCollector);
+    if (schema == null) {
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSource.java]: don't we need to re-fetch schema if config changes?

File: src/test/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Tests to verify configuration of {@link SplunkStreamingSource}
+ * <p>
+ * By default all tests will be skipped, since Splunk credentials are needed.
+ * <p>
+ * Instructions to enable the tests:
```
Comment (mlozbin-cybervisiontech) [inline src/test/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSourceTest.java:60]: should be mentioned in README file

File: widgets/Splunk-streamingsource.json [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+          "label": "Read Timeout (milliseconds)",
+          "name": "readTimeout",
+          "widget-attributes": {
+            "min": "0",
+            "default": "60000"
+          }
+        },
+        {
+          "widget-type": "select",
+          "label": "Number of Retries",
```
Comment (mlozbin-cybervisiontech) [inline widgets/Splunk-streamingsource.json:167]: why not a number?

**Review by mlozbin-cybervisiontech — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java (hunk 1 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:83]: Done.

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSource.java (hunk 2 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSource.java]: Done.

File: src/test/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSourceTest.java (hunk 3 above)
Comment (rostyslavmorozov) [reply src/test/java/io/cdap/plugin/splunk/source/streaming/SplunkStreamingSourceTest.java:60]: Done.

File: widgets/Splunk-streamingsource.json (hunk 4 above)
Comment (rostyslavmorozov) [reply widgets/Splunk-streamingsource.json:167]: It is a requirement from design doc.

File: src/main/java/io/cdap/plugin/splunk/source/batch/SplunkBatchSource.java [hunk 5]
```diff
@@ -47,11 +48,11 @@
 
   public static final String NAME = "Splunk";
 
-  private final SplunkBatchSourceConfig config;
+  private final SplunkSourceConfig config;
   private Schema schema;
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/splunk/source/batch/SplunkBatchSource.java]: this field does not make sense any more

File: README.md [hunk 6]
```diff
@@ -1,2 +1,22 @@
...
+By default all tests will be skipped, since Splunk credentials are needed.
+
+Instructions to enable the tests:
+ 1. Create/use existing Splunk account.
+ 2. Create HEC Token for writing with separate index for testing.
+ 3. Create API Token for reading.
+ 4. Run the tests using the command below:
+
+```
+mvn clean test -Dsplunk.test.token.hec= -Dsplunk.test.token.api= -Dsplunk.test.url.write= -Dsplunk.test.url.read= -Dsplunk.test.index=
```
Comment (mlozbin-cybervisiontech) [inline README.md:21]: please add descriptions for all system properties

**Review by mlozbin-cybervisiontech — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/splunk/source/batch/SplunkBatchSource.java (hunk 5 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/splunk/source/batch/SplunkBatchSource.java]: Removed.

File: README.md (hunk 6 above)
Comment (rostyslavmorozov) [reply README.md:21]: Done.

**Pull request closed without merging**